src/wasm-lib/kcl/src/execution/mod.rs

```diff
@@ -496,17 +496,55 @@ impl ExecutorContext {
     pub async fn run_mock(
         &self,
         program: crate::Program,
-        program_memory_override: Option<ProgramMemory>,
+        use_prev_memory: bool,
+        variables: IndexMap<String, KclValue>,
     ) -> Result<ExecOutcome, KclErrorWithOutputs> {
         assert!(self.is_mock());
 
         let mut exec_state = ExecState::new(&self.settings);
-        if let Some(program_memory_override) = program_memory_override {
-            exec_state.mod_local.memory = program_memory_override;
+        let mut mem = if use_prev_memory {
+            cache::read_old_memory()
+                .await
+                .unwrap_or_else(|| exec_state.memory().clone())
+        } else {
+            exec_state.memory().clone()
+        };
+
+        // Add any extra variables to memory
+        let mut to_restore = Vec::new();
+        for (k, v) in variables {
+            crate::log::log(format!("add var: {k}"));
+            to_restore.push((k.clone(), mem.get(&k, SourceRange::default()).ok().cloned()));
+            mem.add(k, v, SourceRange::synthetic())
+                .map_err(KclErrorWithOutputs::no_outputs)?;
         }
 
+        // Push a scope so that old variables can be overwritten (since we might be re-executing some
+        // part of the scene).
+        mem.push_new_env_for_scope();
+
+        *exec_state.mut_memory() = mem;
+
         self.inner_run(&program.ast, &mut exec_state).await?;
-        Ok(exec_state.to_wasm_outcome())
+
+        // Restore any temporary variables, then save any newly created variables back to
+        // memory in case another run wants to use them.
```

Comment:
Can you explain the code below here a bit more? I thought we didn't cache anything from mock execution in the execution cache. Is it because this branch avoids sending program memory to TS, so we always need to keep the memory from the last run? In `inner_run()`, there's `if not mock { cache::write_old_memory() }`. But in `run_mock()`, we always call `cache::write_old_memory()`. Seems strange.